dynamiqs/qarrays/sparse_dia_qarray.py

```diff
@@ -247,10 +258,40 @@ def _matmul_dia(self, other: SparseDIAQArray) -> QArray:
 
         return SparseDIAQArray(self.dims, tuple(out_offsets), jnp.vstack(out_diags))
 
-    def __and__(self, y: QArray) -> QArray:
+    def __pow__(self, y: Scalar) -> QArray:
         return NotImplemented
 
-    def __pow__(self, y: Scalar) -> QArray:
+    def _kronecker_dia(self, other: SparseDIAQArray) -> SparseDIAQArray:
+        N = other.diags.shape[-1]
+        out_offsets = jnp.ravel(
+            jnp.asarray(self.offsets) * N + jnp.asarray(other.offsets)[:, None],
+            order='F',
+        )
+        out_diags = jnp.kron(self.diags, other.diags)
+        return tuple(out_offsets), out_diags
+
+    def _clean_kronecker_dia(
+        self, offsets: tuple[int, ...], diags: ArrayLike
+    ) -> SparseDIAQArray:
+        offsets = jnp.sort(jnp.asarray(offsets))
+        unique_offsets, inverse_indices = jnp.unique(offsets, return_inverse=True)
+        count = unique_offsets.shape[0]
+        out_diags = jnp.zeros((count, diags.shape[-1]))
+        for i in range(count):
+            mask = inverse_indices == i
+            out_diags = out_diags.at[i, :].set(jnp.sum(diags[mask, :], axis=0))
+        return SparseDIAQArray(
+            tuple(o.item() for o in unique_offsets), out_diags, self.dims
+        )
+
+    def __and__(self, other: QArrayLike) -> SparseDIAQArray | Array:
+        if isinstance(other, SparseDIAQArray):
+            temp_o, temp_d = self._kronecker_dia(other=other)
+            return self._clean_kronecker_dia(temp_o, temp_d)
+
+        elif isinstance(other, get_args(QArrayLike)):
+            return jnp.kron(self.to_dense(), other)
```

Comment:
This is still not fixed: should output a `DenseQArray` and not an `Array` (cf my previous comment).